### PR TITLE
IntelliJ IDEA: Change Setup Instructions #7322

### LINKED ISSUE #6702: Dev docs: tweaks Intellij instructions
Some minor things I noticed while dogfooding: [to be updated as I dogfood more]

> `Configure → Project Defaults → Project Structure.`

Windows version of Intellij doesn't seem to have this menu path

> Click `Import existing project` and select the local repository folder

This menu path is not present in the main UI

image

Initially I thought I should use `...` button, but it didn't work (I was able to choose the project folder but the OK button remained disabled). Then I tried using the button with the folder icon and itworked.

> Go to Run → Play
Use `Run...` instead of `Play`?

> Click + → .... Select the App Engine SDK you downloaded in Step 3 of the Setting up a development 
> environment guide.

If the user chose the `appengine-sdk` folder instead of the sub-folder inside it (note that `appengine-sdk` is the folder mentioned in _Step 3_) it will result in this error at a later stage:
`Error running Dev Server: 'C:/Users/damithch/.gradle/appengine-sdk' isn't valid App Engine SDK installation: 'C:\Users\damithch\.gradle\appengine-sdk\lib\appengine-tools-api.jar' not found`

### LINKED ISSUE #6971: Intellij: output directory setting does not persist
OS: Windows
Problem: this setting get reset to the `.war` file after a while (screenshot shows the correct setting). I'm not sure what exactly causes the change of the setting (may be restarting the IDE). When not set correctly, it can cause '... does not seem to be enhanced` error.

image

### LINKED ISSUE #6972: Intellij: dev server data wiped off periodically
OS: Windows

Problem:
Dev server data gets wiped off after a while. Not sure what action causes it though, probably the `build` action.
This is not the case with Eclipse. The data persist unless deleted manually.

### LINKED ISSUE #7158: Intellij: Fix CI tests failing due to duplicate files
'CI Tests' run configuration fails with the following error:

```
org.testng.TestNGException: 
An error occurred while instantiating class teammates.test.cases.logic.StudentsLogicTest: null
	at org.testng.internal.ClassHelper.createInstance1(ClassHelper.java:401)
	at org.testng.internal.ClassHelper.createInstance(ClassHelper.java:302)
	at org.testng.internal.ClassImpl.getDefaultInstance(ClassImpl.java:118)
	at org.testng.internal.ClassImpl.getInstances(ClassImpl.java:183)
	at org.testng.internal.TestNGClassFinder.<init>(TestNGClassFinder.java:128)
	at org.testng.TestRunner.initMethods(TestRunner.java:413)
	at org.testng.TestRunner.init(TestRunner.java:239)
	at org.testng.TestRunner.init(TestRunner.java:209)
	at org.testng.TestRunner.<init>(TestRunner.java:157)
	at org.testng.SuiteRunner$DefaultTestRunnerFactory.newTestRunner(SuiteRunner.java:546)
	at org.testng.SuiteRunner.init(SuiteRunner.java:159)
	at org.testng.SuiteRunner.<init>(SuiteRunner.java:113)
	at org.testng.TestNG.createSuiteRunner(TestNG.java:1274)
	at org.testng.TestNG.createSuiteRunners(TestNG.java:1261)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1115)
	at org.testng.TestNG.run(TestNG.java:1032)
	at org.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:72)
	at org.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:127)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
Caused by: java.lang.ExceptionInInitializerError
	at teammates.logic.core.AccountsLogic.<clinit>(AccountsLogic.java:32)
	at teammates.logic.api.Logic.<clinit>(Logic.java:68)
	at teammates.test.cases.BaseComponentTestCase.<clinit>(BaseComponentTestCase.java:34)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
	at org.testng.internal.ObjectFactoryImpl.newInstance(ObjectFactoryImpl.java:29)
	at org.testng.internal.ClassHelper.createInstance1(ClassHelper.java:389)
	... 22 more
Caused by: javax.jdo.JDOFatalUserException: Duplicate PMF name "transactions-optional" found in file:/C:/TEAMMATES-sandbox/teammates-intellij/build/resources/main/META-INF/jdoconfig.xml and file:/C:/TEAMMATES-sandbox/teammates-intellij/out/production/teammates-intellij/META-INF/jdoconfig.xml.
	at javax.jdo.JDOHelper.getNamedPMFProperties(JDOHelper.java:1300)
	at javax.jdo.JDOHelper.getPropertiesFromJdoconfig(JDOHelper.java:1232)
	at javax.jdo.JDOHelper.getPersistenceManagerFactory(JDOHelper.java:1079)
	at javax.jdo.JDOHelper.getPersistenceManagerFactory(JDOHelper.java:914)
	at teammates.storage.api.EntitiesDb.<clinit>(EntitiesDb.java:50)
	... 31 more
```

### LINKED ISSUE #7288: Add stylelint support to IntelliJ
Stylelint support is built into IntelliJ, instructions can be updated to show how to setup.

In addition, gradle task(s) can be created to automatically setup static analysis tools such as stylelint.

### LINKED ISSUE #7322: IntelliJ IDEA: Change Setup Instructions
The current setup causes some issues.

**Issues**
#6972 
Not sure how to reproduce this but should not happen, possibly due to conflict with sharing build folder with Gradle. Output folder in IntelliJ to be modified to avoid the problem altogether. Instructions to be added on changing the datastore path as well so the datastore does not get wiped on clean or rebuild.

#6919 (Similar issue here: #7022)
Mystery problem that I'm guessing more likely due to user error. My suggestion is to remove the troubleshooting guide for that so we can re-troubleshoot the problem if they occur again.

#6860 
To resolve by #6980 for manual setup.
To resolve by making user open as an IDEA project instead of importing as a gradle project in automated setup.

#6702
Already partially addressed by #6764, #6809. New instructions will address the rest of the issues.

#7158
Probably caused due to IntelliJ building the files into two folders, duplicating them in `build` and `out` folder

#6971
To resolve by changing build directory in IntelliJ to be different from Gradle and separate enhancer task for IntelliJ.

### PR DESCRIPTION
**Pull request by LiHaoTan**
Closes #7322

**Checklist**

- [x] Automated Setup Docs
- [x] Manual Setup Docs
- [x] Automated Setup Behavior Docs
- [x] Depends on #7361 to be merged
- [x] Revert #6918

**Outline of Solution**
1. Use single module instead of multiple modules, so only one build directory is used.
2. Add automated setup for IntelliJ.
3. Change instructions for manual setup.
3. Add IntelliJ specific Gradle task for JDO enhancement.

**Consequent issues closed/fixed**
Fixes #6972 by changing the build output path to differ from Gradle. In addition, instructions are added for the user to put their datastore path in another folder. More info [here](https://github.com/TEAMMATES/teammates/pull/7354/commits/c871380887d591bf1269c38252213a9e48df41a8)

#6919 was resolved previously but because I could not reach the original contributor it becomes a mystery. I do believe it to be a user error though because the instructions are easy to miss. I have not thought of a good way to address that in the manual setup but the automated setup would not have such an issue.

Thus [docs/troubleshooting-guide.md](https://github.com/TEAMMATES/teammates/pull/6918/files#diff-7bb9f7f577d1c74e5d6259ba1d063058) is reverted even though I was not able to confirm the cause because the solution is not right. There is nothing wrong with the `Output directory` path even if the cause is (which is unlikely unless there is an IntelliJ bug). The path may also not persist because the Gradle build is the "source of truth" and so the manual setting can get reset anytime without warning. More details in 1e65b66. We can also re-troubleshoot on a case by case basis if such an issue were to occur again so we actually find the cause.

#6860 was closed by the user when the issue was not resolved. However, it was fixed by #6980. The new instructions here takes a different approach in the automated setup but the same approach as #6980 in the manual setup.

Fixes #6702
* Configure → Project Defaults → Project Structure
* Click Import existing project and select the local repository folder
> Both of the instructions forces the user to be on the welcome screen and force defaults on the user they may not want. The new instructions do not depend on the user being on the welcome screen.

This menu path is not present in the main UI
image
> assemble is no longer required in the new instructions

* Go to Run → Play
Use Run... instead of Play?
> Already closed by #6764

* Click + → .... Select the App Engine SDK you downloaded in Step 3 of the Setting up a development
environment guide.
> Already closed by #6809

Fixes #7158 by having only one build folder. Previously there was `out` and `build` which can cause the error. Issue further described in 	1e65b66.

Fixes #6971
There are two issues at play here.
1. Build directory is wrong. `IntelliJ` and `Gradle` (or rather `appengine` because Gradle does not explode wars by default) actually uses different exploded war directories. Further details in 1e65b66.
2. In some sense, because a `Gradle` task is used to enhance the classes instead of using IntelliJ Google App Engine integration, the directory for the exploded war need to match the one in the `appengine` plugin. However, it is described in 1e65b66 the problems with that.

Instead of using a `Gradle` task to enhance, it is better to use the Google App Engine integration in IntelliJ to perform the enhancement but it is not working properly, thus a new `Gradle` task specifically for enhancing in IntelliJ is created. I was not able to find out why the enhancement in Google App Engine integration was only working partially but it is not worth to investigate further due to the upcoming `Objectify` migration, so it is to be noted this is a temporary approach.

**Dependent issue closed**
Closes #7288 as the automated setup sets up Stylelint automatically. Thus manual instructions for setting up Styleint is also added. While issue #7288 could be separated and addressed with a different PR there is coupling between the setup instructions and the issue so it will require additional unneeded effort to separate them.

**Caveats**
* Single module instead of multiple modules for the project is used.

   While multiple modules has benefits in separating the dependencies, I was not able to come up with a good solution to support it.

    According to Jetbrains there are also additional benefits in that multiple modules per source set solves some [problems](https://blog.jetbrains.com/idea/2016/03/intellij-idea-2016-1-is-here/). However, multiple modules per source set also caused a number of [problems](https://stackoverflow.com/questions/36330542/control-automatic-creation-of-idea-modules-from-gradle-sourcesets-in-2016-1).

   Two solutions were considered for using multiple modules:
   1. Generating project files through the `idea`plugin in Gradle. However, the plugin is still lacking and [does not yet support](https://goo.gl/tx5lJ2) generating the better directory-based project.
  2. Editing the project files after they are created by the IntelliJ. However, I was not able to come up with a clean solution for this. See 1e65b66 and b113b12.
* IntelliJ JDO enhancement uses an unclean approach. (Will be resolved after Objectify migration since enhancement is no longer needed)
* ES6 Javascript files are not transpiled to ES5 in IntelliJ automatically.
* Include in project build is easy to miss if they did not read [docs/development.md](https://github.com/TEAMMATES/teammates/blob/master/docs/development.md#managing-the-dev-server).

### CODE REVIEW

**Comment by LiHaoTan** [comment]
Ready for review

**Comment by damithc** [comment]
Tried the instructions. Generally went well. Some nits:

1. add `as follows:` here?. At first it wasn't clear to me that the instructions are given below.
image

2. `Gradle VM Options` should be in code format?
image

3. Got this warning when opening the newly-created project in Intellij. Not sure it's due to my previous setups. Anyway, the right response seems to be 'remove'
image

4. Step 2 _must_ be done before step 3? Also, `setupIntellijSettings` --> `setupIntellijProjectSettings`?
image

**Comment by LiHaoTan** [comment]
1. Added `as follows:` to prerequisites.
image

2. Added the code formatting.
image

3. I think it should be due to the previous setups, I think IntelliJ ended up duplicating the modules and they detected one is rendered obsolete.
image

4. It should be in that order because I cannot configure some settings as they are user specific until they open the project, but since you already had a project configured previously, following the steps in the order of `1, 3, 2` would work as well.
image

Because there is a previous setup already done, I am not sure if the instructions have worked correctly or incorrectly. Do you mind if you do a fresh project setup(just delete the .`idea` folder) ?

**Comment by damithc** [comment]
I did the setup twice, second time with a fresh clone. Strangely, the warning (item 3) popped up both times. Will be doing another set up in another computer soon. That computer doesn't have any previous setups. Will let you know if something noteworthy happens during that set up.

**Comment by damithc** [comment]
I tried to do a fresh setup in a new computer. The warning in item 3 still showed up. May be because my root directory is not `teammates`? I did not observe any other problems.

**Comment by LiHaoTan** [comment]
@damithc ah I did not consider a different project root. The current setup will cause some minor quirks if you used a different project root.

I added some settings that will now support a different project root.

If you refresh your project, the module with the name of `$ProjectName$` should now be asked to be removed, and you would have to re-configure the `teammates` module, you would probably receive some pop-ups like in the manual setup.

Alternatively, you can do a fresh setup and should no longer get any pop-ups.

**Comment by whipermr5** [comment]
@LiHaoTan Sorry for the delay! Reviewing now :)

File: docs/ide-usage.md [hunk 1]
```diff
@@ -86,6 +86,12 @@ The following plugins are needed:
 Supported IntelliJ versions: IntelliJ IDEA Ultimate Edition (required to work with Google App Engine).
 You can sign up for the free [JetBrains student license](https://www.jetbrains.com/student/) if you are a student registered in an educational institution.
 
+**NOTE**
+> If you are migrating from Eclipse, you should delete `.project`, `.classpath`, `.launches`,
+> `src/main/webapp/WEB-INF/classes`, `src/main/webapp/WEB-INF/lib` and `src/main/webapp/WEB-INF/appengine-generated`.\
+> You can also copy `local_db.bin` from `src/main/webapp/WEB-INF/appengine-generated` and copy it to the appropriate
```
Comment (whipermr5) [inline docs/ide-usage.md]: ~~and copy it~~

File: docs/ide-usage.md [hunk 2]
```diff
@@ -86,27 +86,118 @@ The following plugins are needed:
...
-   * Click `Import project` and select the local repository folder.
-   * Click `Import project from external model → Gradle`.
-   * Click `Next`.
-   * Check `Use auto-import`. Ensure that `1.7` is used for the `Gradle JVM`.
-   * Click `Finish`.
+**NOTE**
+> If you are migrating from Eclipse, you should delete `.project`, `.classpath`, `.launches`,
+> `src/main/webapp/WEB-INF/classes`, `src/main/webapp/WEB-INF/lib` and `src/main/webapp/WEB-INF/appengine-generated`.\
+> You can also copy `local_db.bin` from `src/main/webapp/WEB-INF/appengine-generated` and copy it to the appropriate
+> location your dev server in IntelliJ reads the datastore from after your IntelliJ setup.
```
Comment (whipermr5) [inline docs/ide-usage.md]: Can this location be made explicit?

File: docs/ide-usage.md [hunk 3]
```diff
@@ -86,27 +86,118 @@ The following plugins are needed:
...
+   1. If you used a different project root directory other than `teammates`, you will receive a dialog box with the
+      message `The modules below are not imported from Gradle anymore. Check those to be removed from the ide project too:`.\
+      You should see a module with your project root directory name being checked. Click `OK`.
+   1. Go to `File → Project Structure... → Modules`.
+      Click on the `teammates` module, then under `Sources`, click on the `build` folder and click `Excluded` and then `OK`.
+   1. You should see a dialog box with the message:\
+      `Frameworks detected: Google App Engine, Web, JPA frameworks are detected in the project`.\
+      **OR**\
+      `Frameworks detected: OSGi, Google App Engine, Web, JPA frameworks are detected in the project`.\
+      Click `Configure`, ensure that there is only `Google App Engine` and `JPA` frameworks being shown, otherwise make
```
Comment (whipermr5) [inline docs/ide-usage.md]: Click `Configure`~~,~~ **and** ensure that ~~there is~~ only `Google App Engine` and `JPA` frameworks ~~being~~ **are** shown,

File: docs/ide-usage.md (hunk 1 above)
Comment (damithc) [inline docs/ide-usage.md]: On a related note, I wish intellij/gradle can preserve the local db through builds/cleans. Test data previous entered should not be wiped out during builds. But probably no need to do in this PR.

File: docs/staticAnalysis.md [hunk 4]
```diff
@@ -209,17 +282,37 @@ The link to the report will be displayed in each PR, or by clicking the badge on
...
+
+### IntelliJ IDEA
+
+For Java tests, you can measure code coverage for the project using `Run → Run... → CI Tests → ▶ → Cover `.
+
+**NOTE**
+> There are some packages and classes that are excluded when using Jacoco on the comamnd line which are not excluded
+when you run them in IntelliJ, thus the coverage statistics you see in IntelliJ may not match what you see on `Codecov`. 
+
+Alternatively, you can right click a class with TestNG test(s), and click `Run '$FileClass$' with Coverage`, this will use
```
Comment (whipermr5) [inline docs/staticAnalysis.md]: click a class with TestNG test(s)~~,~~ and click

File: docs/staticAnalysis.md [hunk 5]
```diff
@@ -209,17 +282,37 @@ The link to the report will be displayed in each PR, or by clicking the badge on
...
 A coverage lower bound is enforced via `AllJsTests.java`, lower than which the build will fail.
 
-### Local build
-
-For Java tests, choose `Coverage as TestNG Test` instead of the usual `Run as TestNG Test` to run the specified test or test suite.
-The coverage will be reported in Eclipse after the test run is over.
+### CLI
 
-Alternatively, use Gradle to run the tests, and obtain the coverage data with `jacocoReport` task, i.e:
+You can use use Gradle to run tests, and obtain the coverage data with `jacocoReport` task, i.e:
```
Comment (whipermr5) [inline docs/staticAnalysis.md]: run tests~~,~~ and

File: docs/staticAnalysis.md [hunk 6]
```diff
@@ -199,6 +234,44 @@ To run all static analysis tasks in one sitting, run the following two commands:
...
+If you wish to run inspections for the whole project, you can do `Analyze → Inspect Code... → Whole project`. You may
+also wish to learn more about code inspections by referring to
+[IntelliJ IDEA's documentation](https://www.jetbrains.com/help/idea/2017.1/code-inspection.html).
+
+**NOTE**
+> `FindBugs` will only appear in the inspection results if you ran it manually before running
+> `Analyze → Inspect Code... → Whole project`.
+
+`PMD` is provided as a plugin and do not run automatically. You can run it by selecting a number of files, right clicking,
+select `Run PMD` and then the option you want.
```
Comment (whipermr5) [inline docs/staticAnalysis.md]: select**ing*** `Run PMD` and then **choosing** the option you want.

File: docs/ide-usage.md (hunk 1 above)
Comment (whipermr5) [inline docs/ide-usage.md]: @LiHaoTan has included instructions how to fix this here: https://github.com/TEAMMATES/teammates/pull/7354/files#diff-1bf07101845de6da98f1c948b82c5ce0R84

**Pull request merged**